### PR TITLE
New MemoryAdapter

### PR DESCRIPTION
**Pull request by Maronato**
## ☕️ Reasoning

This PR implements a new `MemoryAdapter` that is useful for testing and debugging by requiring no setup and keeping all it's data in memory.

## 🧢 Checklist

- [ ] Documentation
- [ ] Tests
- [ ] Ready to be merged

## 🎫 Affected issues

Related to #8808 

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

### CODE REVIEW

**Comment by ThangHuuVu** [comment]
hi @Maronato could you extract the memory adapter and make a new PR for it? I think we should merge it first, we can use it for testing and stuff, the passkey related code can be merged afterward.

**Comment by ndom91** [comment]
@Maronato are you still interested in getting these merged? We'd love to move forward on the two PRs. 

If so, could you checkout the merge conflicts and also add some small docs in typedoc format at the top of `packages/adapter-memory/src/index.ts` just saying its mainly for testing? (see: [prisma adapter typedoc docs](https://github.com/nextauthjs/next-auth/blob/main/packages/adapter-prisma/src/index.ts) as example) :pray:

**Comment by Maronato** [comment]
yup, will work on the conflicts and your suggestions this weekend

File: packages/adapter-memory/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+  const characters =
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+  const charactersLength = characters.length
+  let counter = 0
+  while (counter < length) {
+    result += characters.charAt(Math.floor(Math.random() * charactersLength))
+    counter += 1
+  }
+  return result
+}
```
[the remark this answers is not in the document]
Comment (Maronato) [reply packages/adapter-memory/src/index.ts:246]: While I agree, I am getting `ReferenceError: crypto is not defined` when running the tests.
`node -v` gives `18.18.0` which is the default for the project, and the tests are run with `pnpm test --filter=@auth/memory-adapter`

I'm not gonna spend too much time debugging this, so I'll stick to my custom makeid function for now. What the function does is fairly simple - it picks random letters from `characters` and builds a word of size `length`.
I'm going to add some comments nonetheless to make it clearer.

**Comment by Maronato** [comment]
@ThangHuuVu @ndom91  Added docs and removed passkey stuff.
I'll add the passkey tests to the passkey PR instead once this is merged.

**Comment by Maronato** [comment]
Due to recent changes in Nextjs, servers are no longer persisted in memory between requests. This makes it impossible to persist the adapter state, which breaks this implementation. Closing.

**Pull request closed without merging**